File: python/flower-field/flower_field.py
def annotate(garden):
    def process_square(row, col):
        value = garden[row][col]
        if value == '*':
            return '*'
        if value == ' ':
            count = count_adjacent(row, col)
            return str(count) if count > 0 else ' '
        raise ValueError('The board is invalid with current input.')
    def count_adjacent(row, col):
        adj_squares = [
            (row - 1, col - 1),
            (row - 1, col),
            (row - 1, col + 1),
            (row, col - 1),
            (row, col + 1),
            (row + 1, col - 1),
            (row + 1, col),
            (row + 1, col + 1),
        ]
        neighbors = [garden[r][c] for r, c in adj_squares
                     if 0 <= r < rows and 0 <= c < cols]
        return len([adj for adj in neighbors if adj == '*'])
    rows = len(garden)
    if rows > 0:
        cols = len(garden[0])
    else:
        return []
    if any([len(r) != cols for r in garden]):
        raise ValueError('The board is invalid with current input.')
    return [''.join([process_square(row, col) for col in range(cols)])
            for row in range(rows)]

File: julia/flower-field/flower-field.jl
function annotate(flowerfield)
    function process_square(row, col)
        value = flowerfield[row][col]
        if value == '*'
            return "*"
        end
        if value == ' '
            count = count_adjacent(row, col)
            return  count > 0 ? string(count) : " "
        end
    end

    function count_adjacent(row, col)
        offsets = [(r + c * 1im) for r = -1:1 for c = -1:1]
        adj_squares = [row + col * 1im + offset for offset in offsets]
        neighbors = [flowerfield[real(sq)][imag(sq)] for sq in adj_squares
                     if 1 <= real(sq) <= rows && 1 <= imag(sq) <= cols]
        return length([adj for adj in neighbors if adj == '*'])
    end

    rows = length(flowerfield)
    if rows > 0
        cols = length(flowerfield[1])
    else
        return []
    end

    [join([process_square(row, col) for col in 1:cols]) for row in 1:rows]
end
